[Rui Yan] iP

chowRuiYan opened the pull request:
# Dukey
> If you have a task, I will handle it for you!
> -Dukey

Dukey is a Todo-List that helps manage your tasks. Here are some basic features:
* Add ~~task~~ different type of tasks like (Todo, Deadline and Event)
* Mark tasks as done
* Delete tasks
* Find tasks with keywords
* Display all the tasks you have

All you need to do is,
1. Download it from [here](https://github.com/chowRuiYan/ip/releases).
2. Click on **Assets**, then click on **iP.jar**.
3. Run and add your tasks.
4. I'll do the rest for you!

It is *EASY* to use and most importantly, it is **FREE**  :grin:

Future updates:
- [x] Finding tasks with keywords
- [ ] Introducing calendar (coming real soon) 
- [ ] Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's a the `main` method.
```Java
public class Main {
    public static void main(String[] args) {
        Duke dukey = new Duke(filePath);
        dukey.run();
    }
}
```

Commit:
Add Gradle support

chowRuiYan committed:
Add Level-1

chowRuiYan committed:
Add Level-2

chowRuiYan committed:
Add Level-3. Change to OOP.

chowRuiYan committed:
Add Level-4

chowRuiYan committed:
Add input.txt and EXPECTED.txt. Minor changes to Task class in Duke.java.

chowRuiYan committed:
Add Level-5

chowRuiYan committed:
Add Level-6

chowRuiYan committed:
Add Level-7

chowRuiYan committed:
Add Level-8

chowRuiYan committed:
Merge branch 'branch-Level-7'

chowRuiYan committed:
Merge branch-Level-7 and branch-Level-8

chowRuiYan committed:
Add A-MoreOOP

chowRuiYan committed:
Add A-Packages

chowRuiYan committed:
Add A-JUnit

chowRuiYan committed:
Update .gitignore to include MANIFEST.MF

chowRuiYan committed:
Add A-JavaDoc

chowRuiYan committed:
Add A-CodingStandard

chowRuiYan committed:
Add Level-9

chowRuiYan committed:
Merge branch 'branch-A-Javadoc'

chowRuiYan committed:
Merge branch 'branch-A-CodingStandard'

chowRuiYan committed:
Merge branch 'branch-Level-9'

GnohChengYi commented on src/main/java/duke/Command.java:
```diff
@@ -0,0 +1,258 @@
+ */
+public abstract class Command {
+    /**
+     * Abstract method meant to indicate the actions taken for different command.
```
> In method header comments, the first sentence should start in the form Returns ..., Sends ..., Adds ... (not Return or Returnning etc.)

Would it be better this way?

```suggestion
     * Indicates the actions taken for different command.
```

A client who uses this class does not need to know that the method is abstract. A subclass of `Command` would know it is abstract, otherwise the code will not compile.

[Here](https://docs.oracle.com/en/java/javase/11/docs/api/java.desktop/javax/swing/Action.html) is an example from Java 11 API. Notice that `getValue(String key)` is listed under abstract methods, but its Javadoc does not mention anything about it being `abstract`.

chowRuiYan replied:
Many thanks for all your comments, I'll change it accordingly.

GnohChengYi commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,68 @@
+        String[] inputs = input.split(" ");
+        String action = inputs[0];
+
+        switch (action.toLowerCase()) {
```
I like how you made the commands case insensitive 👍

GnohChengYi commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,125 @@
+    };
+
+    /**
+     * UI when a task has been added.
```
> In method header comments, the first sentence should start in the form Returns ..., Sends ..., Adds ... (not Return or Returnning etc.)

Would it be better this way?
```suggestion
     * Displays UI to show a task has been added.
```

I noticed the same issue in several other places within this file.

GnohChengYi commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,125 @@
+     *
+     * @param task The Task that has been marked as done.
+     */
+    public void doneUi(TaskList.Task task) {
```
Another issue is that 

> Names representing methods must be verbs and written in camelCase.

Would it be better to rename the methods in the following way?
```suggestion
    public void displayDoneUi(TaskList.Task task) {
```

This issue applies to several other Ui methods within this file.

GnohChengYi reviewed (commented):
LGTM, almost. Maybe just a few nits to fix.

chowRuiYan committed:
Merge remote-tracking branch 'origin/add-gradle-support'

zhing22 commented on src/main/java/duke/Command.java:
```diff
@@ -0,0 +1,258 @@
+    /**
+     * Handles the adding of new task into Duke.
+     */
+    public static class AddCommand extends Command {
```
In the code quality guidelines [here](https://nus-cs2103-ay2122s1.github.io/website/se-book-adapted/chapters/codeQuality.html#guideline-name-well) under naming section, theres this line that says "Use __nouns__ for classes/variables and verbs for methods/functions". 
Perhaps it would be nice to make your class name a noun (maybe something like ```CommandAdder extends CommandImplementer```? but i'll leave that to you)

zhing22 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,68 @@
+        case "list":
+            return new Command.ListCommand();
+        case "done":
+            if (inputs.length == 1) {
+                throw new DukeException("Insufficient input received! " +
+                        "Please indicate the task number of the completed task.");
+            } else {
+                try {
+                    int index = Integer.parseInt(inputs[1]);
+
+                    return new Command.DoneCommand(index);
+                } catch (NumberFormatException e) {
+                    throw new DukeException("Invalid input! Please enter the task number after 'done'.");
+                }
+            }
```
Perhaps it would be nice to reduce nested coding. Here I removed the else block so the try-catch block is at the same indentation as the other lines in the ```done``` block. 

```suggestion
            if (inputs.length == 1) {
                throw new DukeException("Insufficient input received! " +
                        "Please indicate the task number of the completed task.");
            } 
            
            try {
                int index = Integer.parseInt(inputs[1]);
                return new Command.DoneCommand(index);
            } catch (NumberFormatException e) {
                throw new DukeException("Invalid input! Please enter the task number after 'done'.");
            }
```

zhing22 reviewed (commented):
Nice code! I really had nothing much to comment. I like how you split your code into multiple subclasses and methods which made it really clear and easy to read! 👍

chowRuiYan committed:
Add A-Gradle.

Add changes according to review given by peer.

chowRuiYan committed:
Add A-Checkstyle.

Add changes to rectify errors according to checkstyle violations.

chowRuiYan committed:
Merge branch 'branch-A-Checkstyle'

chowRuiYan committed:
Add Level-10

chowRuiYan committed:
Merge branch 'branch-Level-10'

chowRuiYan committed:
Add A-Varargs.

Refactored to follow better OOP practice.

chowRuiYan committed:
Add A-Assertions.

chowRuiYan committed:
Add A-CodeQuality.

Parser: extract method for error checking and follow DRY principle.

Similar error checking behavior caused code duplication in parse()
method. Code fragments used to check for error can also be grouped
together.

Unifying all those duplicate codes and extracting them into a method
improves code quality.

chowRuiYan committed:
Merge pull request #1 from chowRuiYan/branch-A-Assertions

Add A-Assertions.

chowRuiYan committed:
Merge branch 'master' into branch-A-CodeQuality

chowRuiYan committed:
Merge pull request #2 from chowRuiYan/branch-A-CodeQuality

Add A-CodeQuality.

chowRuiYan committed:
Add A-CI

Github Actions help with automated tasks during the software
development life-cycle.

Add gradle.yml from [here.](https://github.com/se-edu/duke/blob/full-template/.github/workflows/gradle.yml)

Build and check Duke every time a push/pull request event is triggered.

chowRuiYan committed:
Add C-DetectDuplicate

Previous version of duke allows multiple entries of the same task to be
added. It is redundant to add it into our todo-list.

Add method to check if task already exist in the todo-list before adding
it. Gives a using a warning when trying to add duplicated task.

Minor bug fixed in Parser.java: A Deadline task will be added even when
command to add an Event is given.

chowRuiYan committed:
Merge branch 'branch-C-DetectDuplicates'

chowRuiYan committed:
Add A-BetterGui.

Duke will have a different icon to represent an error and also return
the user's input

chowRuiYan committed:
Incomplete commit from previous commit.

chowRuiYan committed:
Add UG.

Added a screenshot of Duke and updated README.md to give an overview of
Duke.

chowRuiYan committed:
Set theme jekyll-theme-minimal

chowRuiYan committed:
Minor changes to abide to coding standards.

chowRuiYan committed:
Merge branch 'master' of https://github.com/chowRuiYan/ip

chowRuiYan committed:
Change MainWindow.java such that GUI will auto close when ExitCommand
is given.

chowRuiYan committed:
Fix JavaFX dependency
